huff0/_generate/gen.go

```diff
@@ -0,0 +1,366 @@
+	} else {
+		MOVL(addr, tmp.As32()) // tmp = uint32(b.in[b.off:b.off+4])
+		MOVQ(brBitsRead, CX.As64())
+		SHLQ(CX, tmp.As64())
+	}
+
+	ORQ(tmp.As64(), brValue)
+
+	Commentf("exhausted = exhausted || (br%d.off < 4)", id)
+	CMPQ(brOffset, U8(4))
+	SETLT(exhausted.As8L())
+	ORB(exhausted.As8L(), exhausted.As8H())
+	Label("skip_fill" + strconv.Itoa(id))
+
+	val := GP64()
+	Commentf("val0 := br%d.peekTopBits(peekBits)", id)
+	if d.bmi2 {
+		SHRXQ(peekBits, brValue, val.As64()) // val = (value >> peek_bits) & mask
+	} else {
+		MOVQ(brValue, val.As64())
+		MOVQ(peekBits, CX.As64())
+		SHRQ(CX, val.As64()) // val = (value >> peek_bits) & mask
+	}
+
+	Comment("v0 := table[val0&mask]")
+	v := reg.RDX
+	MOVW(Mem{Base: table, Index: val.As64(), Scale: 2}, v.As16())
+
+	Commentf("br%d.advance(uint8(v0.entry)", id)
+	MOVB(v.As8H(), out.As8()) // BL = uint8(v0.entry >> 8)
+
+	MOVBQZX(v.As8(), CX.As64())
+	if d.bmi2 {
+		SHLXQ(v.As64(), brValue, brValue) // value <<= n
+	} else {
+		SHLQ(CX, brValue) // value <<= n
+	}
+
+	ADDQ(CX.As64(), brBitsRead) // bits_read += n
+
+	Commentf("val1 := br%d.peekTopBits(peekBits)", id)
+	if d.bmi2 {
+		SHRXQ(peekBits, brValue, val.As64()) // val = (value >> peek_bits) & mask
+	} else {
+		MOVQ(peekBits, CX.As64())
+		MOVQ(brValue, val.As64())
+		SHRQ(CX, val.As64()) // val = (value >> peek_bits) & mask
+	}
+
+	Comment("v1 := table[val1&mask]")
+	MOVW(Mem{Base: table, Index: val.As64(), Scale: 2}, v.As16()) // tmp - v1
+
+	Commentf("br%d.advance(uint8(v1.entry))", id)
+	MOVB(v.As8H(), out.As8H()) // BH = uint8(v0.entry >> 8)
+
+	MOVBQZX(v.As8(), CX.As64())
+	if d.bmi2 {
+		SHLXQ(v.As64(), brValue, brValue) // value <<= n
+	} else {
+		SHLQ(CX, brValue) // value <<= n
+	}
+
+	ADDQ(CX.As64(), brBitsRead) // bits_read += n
+
+	Comment("these two writes get coalesced")
+	Comment("buf[stream][off] = uint8(v0.entry >> 8)")
+	Comment("buf[stream][off+1] = uint8(v1.entry >> 8)")
+	MOVW(out.As16(), Mem{Base: buffer, Index: off, Scale: 1, Disp: id * buffoff})
+
+	Comment("update the bitrader reader structure")
+	Store(brBitsRead.As8(), br.Field("bitsRead"))
+	Store(brValue, br.Field("value"))
+	Store(brOffset, br.Field("off"))
+}
+
+type decompress4x8bit struct {
+	bmi2 bool
+}
+
+func (d decompress4x8bit) generateProcedure(name string) {
+	Package("github.com/klauspost/compress/huff0")
+	TEXT(name, 0, "func(pbr0, pbr1, pbr2, pbr3 *bitReaderShifted, peekBits uint8, buf *byte, tbl *dEntrySingle) uint8")
+	Doc(name+" is an x86 assembler implementation of Decompress4X when tablelog > 8.decodes a sequence", "")
+	Pragma("noescape")
+
+	out := reg.RAX // Fixed since we need 8H
+
+	off := GP64()
+	XORQ(off, off)
+
+	exhausted := reg.RBX                     // Fixed since we need 8H
+	XORQ(exhausted.As64(), exhausted.As64()) // exhausted = false
+
+	peekBits := GP64()
+	buffer := GP64()
+	table := GP64()
+
+	Comment("Preload values")
+	{
+		Load(Param("peekBits"), peekBits)
+		Load(Param("buf"), buffer)
+		Load(Param("tbl"), table)
+	}
```

Comment:
`Load` returns the register. You can do it like this:

```go
peekBits := Load(Param("peekBits"), GP64())
```

Reply:
Yes, I saw this in the examples. For me having separate allocation and the use is a bit cleaner.